jumpscale/packages/auth/bottle/auth.py

```diff
@@ -0,0 +1,284 @@
+import base64
+import json
+from functools import wraps
+from urllib.parse import urlencode
+
+import nacl
+import requests
+from beaker.middleware import SessionMiddleware
+from bottle import Bottle, abort, redirect, request, response
+from nacl.public import Box
+from nacl.signing import VerifyKey
+
+from jinja2 import Environment, FileSystemLoader, select_autoescape
+from jumpscale.god import j
+
+SESSION_OPTS = {"session.type": "file", "session.data_dir": "./data", "session.auto": True}
+REDIRECT_URL = "https://login.threefold.me"
+CALLBACK_URL = "/auth/3bot_callback"
+LOGIN_URL = "/auth/login"
+
+app = Bottle()
+
+templates_path = j.sals.fs.join_paths(j.sals.fs.dirname(__file__), "templates")
+env = Environment(loader=FileSystemLoader(templates_path), autoescape=select_autoescape(["html", "xml"]))
+
+
+@app.route("/login")
+def login():
+    """List available providers for login and redirect to the selected provider (3bot connect)
+
+    Returns:
+        Renders the template of login page
+    """
+    session = request.environ.get("beaker.session")
+    provider = request.query.get("provider")
+    next_url = request.query.get("next_url", session.get("next_url"))
+
+    if provider and provider == "3bot":
+        state = j.data.idgenerator.chars(20)
+        session["next_url"] = next_url
+        session["state"] = state
+        app_id = request.get_header("host")
+        params = {
+            "state": state,
+            "appid": app_id,
+            "scope": json.dumps({"user": True, "email": True}),
+            "redirecturl": CALLBACK_URL,
+            "publickey": j.core.identity.me.nacl.public_key.encode(encoder=nacl.encoding.Base64Encoder),
+        }
+        params = urlencode(params)
+        return redirect(f"{REDIRECT_URL}?{params}", code=302)
+
+    return env.get_template("login.html").render(providers=[], next_url=next_url)
```

Comment:
also u can do j.tools.jinja2.render_template directly from here..

Reply:
resolved using the env from j.tools.jinja